src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,420 @@
+     *
+     * @return a new {@code ExtractExpand} object
+     */
+    static ExtractExpand extractExpand(Extract ext, byte[] info, int length) {
+        return new ExtractExpand(ext, info, length);
+    }
+
+    /**
+     * Defines the input parameters of an Extract operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    final class Extract implements HKDFParameterSpec {
+
+        // HKDF-Extract(salt, IKM) -> PRK
+        private final List<SecretKey> ikms;
+        private final List<SecretKey> salts;
+
+        private Extract() {
+            this(new ArrayList<>(), new ArrayList<>());
+        }
+
+        private Extract(List<SecretKey> ikms, List<SecretKey> salts) {
+            this.ikms = ikms;
+            this.salts = salts;
+        }
+
+        /**
+         * Returns the unmodifiable {@code List} of initial key material values.
+         *
+         * @return the unmodifiable {@code List} of initial key material values
+         */
+        public List<SecretKey> ikms() {
+            return Collections.unmodifiableList(ikms);
+        }
+
+        /**
+         * Returns the unmodifiable {@code List} of salt values.
+         *
+         * @return the unmodifiable {@code List} of salt values
+         */
+        public List<SecretKey> salts() {
+            return Collections.unmodifiableList(salts);
+        }
+
+    }
+
+    /**
+     * Defines the input parameters of an Expand operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    final class Expand implements HKDFParameterSpec {
+
+        // HKDF-Expand(PRK, info, L) -> OKM
+        private final SecretKey prk;
+        private final byte[] info;
+        private final int length;
+
+        /**
+         * Constructor that may be used to initialize an {@code Expand} object
+         *
+         * @param prk
+         *     the pseudorandom key; may be {@code null}
+         * @param info
+         *     the optional context and application specific information
+         * @param length
+         *     the length of output keying material
+         */
+        private Expand(SecretKey prk, byte[] info, int length) {
+            // a null prk could be indicative of ExtractExpand
+            this.prk = prk;
+            this.info = (info == null) ? null : info.clone();
+            if (length < 1) {
+                throw new IllegalArgumentException("length must be >= 1");
+            }
+            this.length = length;
+        }
+
+        /**
+         * Returns the pseudorandom key.
+         *
+         * @return the pseudorandom key
+         */
+        public SecretKey prk() {
+            return prk;
+        }
+
+        /**
+         * Returns the optional context and application specific information.
+         *
+         * @return the optional context and application specific information
+         */
+        public byte[] info() {
```

Comment:
I think the API docs will need to say that it returns a copy of the context/info, don't want readers to assume that an Expand object is mutable by way to modifying the byte[].